.github/workflows/ci.yaml

```diff
@@ -162,61 +162,83 @@ jobs:
-        # use set -o pipefail to keep error code from gradle, c.f. https://stackoverflow.com/a/37128767
-        ( set -o pipefail; ./gradlew build --parallel --build-cache --info | grep -vE "^Transforming (.*) with (.*)Transform" )
+      - name: Cache Gradle
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/.gradle/caches
+            ~/.gradle/wrapper
+          key: gradle-${{ runner.os }}-${{ hashFiles('**/*.gradle*', '**/gradle-wrapper.properties') }}
+
+      - name: Grant execute permission for gradlew
+        run: |
+          chmod +x ./gradlew
+
+      - name: Build
+        shell: bash
+        run: |
+          # use grep to remove log spam due to jetifier transformation, c.f. https://stackoverflow.com/a/52353446
+          # use set -o pipefail to keep error code from gradle, c.f. https://stackoverflow.com/a/37128767
+          ( set -o pipefail; ./gradlew build --parallel --build-cache --info | grep -vE "^Transforming (.*) with (.*)Transform" )
+
+      - name: Cache AVD
+        uses: actions/cache@v2
+        id: avd-cache
+        with:
+          path: |
+            ~/.android/avd/*
+            ~/.android/adb*
+          key: avd-29
+
+      - name: Create AVD and generate snapshot for caching
+        if: steps.avd-cache.outputs.cache-hit != 'true'
+        uses: reactivecircus/android-emulator-runner@v2
+        with:
+          api-level: 29
+          force-avd-creation: false
+          emulator-options: -no-window -gpu swiftshader_indirect -noaudio -no-boot-anim -camera-back none
+          disable-animations: false
+          working-directory: ./fe2-android
+          script: echo "Generated AVD snapshot for caching."
+
+      - name: Cache SonarCloud packages
+        uses: actions/cache@v2
+        with:
+          path: ~/.sonar/cache
+          key: ${{ runner.os }}-sonar
+          restore-keys: ${{ runner.os }}-sonar
 
-    - name: Unit Tests
-      uses: reactivecircus/android-emulator-runner@v2
-      with:
-        api-level: 29
-        arch: x86_64
-        profile: pixel_2
-        emulator-options: -no-window -gpu swiftshader_indirect -no-snapshot -noaudio -no-boot-anim -camera-back none
-        disable-animations: true
-        working-directory: ./fe2-android
-        script: |
-          ( set -o pipefail; ./gradlew testDebugUnitTest createDebugCoverageReport --parallel --build-cache jacocoTestReport --info | grep -vE "^Transforming (.*) with (.*)Transform" )
-          
-    - name: SonarCloud Analysis
-      env:
-        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-        SONAR_TOKEN: ${{ secrets.SONAR_TOKEN_FE2 }}
-      run: |
-        ./gradlew sonarqube
+      - name: Run Tests
+        uses: reactivecircus/android-emulator-runner@v2
+        with:
+          api-level: 29
+          force-avd-creation: false
+          emulator-options: -no-snapshot-save -no-window -gpu swiftshader_indirect -noaudio -no-boot-anim -camera-back none
```

Comment:
"no window" but "gpu" seems weird.

Reply:
I suspect that it means "run this in headless mode" (no window) , and use "swiftshader_indirect" as GPU ( e.g. https://developer.android.com/studio/run/emulator-acceleration )